src/hub/manual/core/mat.rs

```diff
@@ -83,43 +83,133 @@ impl Mat {
     }
 
     #[inline(always)]
     pub(crate) fn _at_mut<T: DataType>(&mut self, i0: i32) -> Result<&mut T> {
-        self.match_format::<T>().and_then(|_| self.ptr_mut(i0).map(|x| unsafe { &mut *(x as *mut _ as *mut T) }))
+        self.match_format::<T>()?;
+        let size = self.size()?;
+        if i0 >= size.height {
+            return Err(Error::new(
+                core::StsOutOfRange,
+                format!("Given coordinate is out of bounds. \
+                         Mat has size {:?} but was given i0: {}", size, i0)));
+        }
+        unsafe { self.at_mut_unchecked(i0) }
+    }
+
+    /// Like `Mat::at_mut()` but performs no bounds or type checks
+    pub unsafe fn at_mut_unchecked<T: DataType>(&mut self, i0: i32) -> Result<&mut T> {
+        self.ptr_mut(i0).map(|x| unsafe { &mut *(x as *mut _ as *mut T) })
     }
 
     #[inline(always)]
     pub(crate) fn _at_2d<T: DataType>(&self, row: i32, col: i32) -> Result<&T> {
-        self.match_format::<T>().and_then(|_| self.ptr_2d(row, col).map(|x| unsafe { &*(x as *const _ as *const T) }))
+        self.match_format::<T>()?;
+        let size = self.size()?;
+        if row >= size.height || col >= size.width {
+            return Err(Error::new(
+                core::StsOutOfRange,
+                format!("Given coordinate is out of bounds. \
+                         Mat has size {:?} but was given row: {}, col: {}", size, row, col)));
+        }
+        unsafe { self.at_2d_unchecked(row, col) }
+    }
+
+    /// Like `Mat::at_2d()` but performs no bounds or type checks
+    pub unsafe fn at_2d_unchecked<T: DataType>(&self, row: i32, col: i32) -> Result<&T> {
+        self.ptr_2d(row, col).map(|x| unsafe { &*(x as *const _ as *const T) })
     }
 
     #[inline(always)]
     pub(crate) fn _at_2d_mut<T: DataType>(&mut self, row: i32, col: i32) -> Result<&mut T> {
-        self.match_format::<T>().and_then(|_| self.ptr_2d_mut(row, col).map(|x| unsafe { &mut *(x as *mut _ as *mut T) }))
+        self.match_format::<T>()?;
+        let size = self.size()?;
+        if row >= size.height || col >= size.width {
+            return Err(Error::new(
+                core::StsOutOfRange,
+                format!("Given coordinate is out of bounds. \
+                         Mat has size {:?} but was given row: {}, col: {}", size, row, col)));
+        }
+        unsafe { self.at_2d_mut_unchecked(row, col) }
+    }
+
+    /// Like `Mat::at_2d_mut()` but performs no bounds or type checks
+    pub unsafe fn at_2d_mut_unchecked<T: DataType>(&mut self, row: i32, col: i32) -> Result<&mut T> {
+        self.ptr_2d_mut(row, col).map(|x| unsafe { &mut *(x as *mut _ as *mut T) })
     }
 
+    /// Note that since the bindings are set up to wrap all Mat sizes in our custom `Size` struct
+    /// instead of a `MatSize` smart pointer, we don't have a good way to check bounds on Mats with
+    /// more than 2 dimensions, so we can't perform a real bounds check here.
```

Comment:
I think we would have to make `Size` an actual `MatSize` instead to perform a check here. In general it seems like >2d matrices are not well supported by OpenCV, so maybe it's acceptable to have this discrepancy.

That said, I wonder if it's worth marking the `at_3d*` methods as unsafe to reflect the fact that they don't perform any bounds checks, since with these new changes it'd be reasonable to extrapolate that the marked-safe `at_3d()` is actually safe like the `at_2d()` and `at()` functions.